### PR TITLE
Add External GUI and secure password with (BoincTasks)

### PR DESCRIPTION
**Pull request by efmer**
commit d78bb769a71ce6167ae65332fa13890a660a2c9b
Author: efmer <boinc@efmer.eu>
Date:   Thu Oct 22 13:57:05 2020 +0200

    Added allowed WiFi IP addresses to connect.

commit c3855c6604bcaa585cff424c00572ceaceb872b7
Author: efmer <boinc@efmer.eu>
Date:   Mon Oct 19 16:45:09 2020 +0200

    Add a way to close the app and service, by pressing back

commit 6a17414292eb783a01965c5bac81aa8e5dad6832
Author: efmer <boinc@efmer.eu>
Date:   Sun Oct 18 08:58:13 2020 +0200

    New module versions

commit bb119b30ba0913d2f464e213adc5e2b62e8c5a9c
Author: efmer <boinc@efmer.eu>
Date:   Sun Oct 18 08:56:48 2020 +0200

    Set Foreground Service

commit 015e26831053e49c4fe1d932d694a60b5b409236
Author: efmer <boinc@efmer.eu>
Date:   Sun Oct 18 08:56:07 2020 +0200

    Updated libraries

commit 2a5cf28cd0c8eead2c2c9bc0fed1dba3b5b769c5
Author: efmer <boinc@efmer.eu>
Date:   Sat Oct 17 18:56:44 2020 +0200

    targetSdkVersion must be 29 for store approval

commit 88bb250a3940e917b50fcfc4d3bbf999cc9e4f64
Merge: 03a069bbbf 16d106128c
Author: efmer <boinc@efmer.eu>
Date:   Sat Oct 17 18:48:38 2020 +0200

    Merge remote-tracking branch 'origin/master' into master

commit 03a069bbbfaf18692a17d8c5902d99539262591d
Author: efmer <boinc@efmer.eu>
Date:   Wed Oct 14 14:17:48 2020 +0200

    Simplify passing data to thread

commit a90426d9e7be9e91fe2a9ec0965bc03d73630916
Author: efmer <boinc@efmer.eu>
Date:   Wed Oct 14 14:15:22 2020 +0200

    The client crashes

commit 4d7526414a5a2eb62d69284d2829418a93cb2a2b
Merge: 9c90ee95e1 418cc0da73
Author: efmer <boinc@efmer.eu>
Date:   Wed Oct 14 08:48:40 2020 +0200

    Merge remote-tracking branch 'origin/master' into master

    # Conflicts:
    #	android/BOINC/build.gradle

commit 9c90ee95e1c3f647d5cfe0b46b9b784cbe672a7d
Author: efmer <boinc@efmer.eu>
Date:   Tue Oct 13 15:09:25 2020 +0200

    Password mismatch in MD5

commit 9fa069693d350bd052fda746fd652854b3756533
Author: efmer <boinc@efmer.eu>
Date:   Mon Oct 12 16:58:56 2020 +0200

    Added AES encrypted password

commit c7d7d8ae3d22bb238079efb6dca8a98685891f69
Merge: b5fe0c2fa2 3cf5fb5e54
Author: efmer <boinc@efmer.eu>
Date:   Sun Oct 11 19:03:09 2020 +0200

    Merge remote-tracking branch 'origin/master' into master

commit b5fe0c2fa28a9ef1fdd6fa11856af2fe7bf44cc4
Author: efmer <boinc@efmer.eu>
Date:   Wed Oct 7 15:43:34 2020 +0200

    Limit connections to WiFi only

commit d9c26c58a526fc2516d1d60ee2d3e658b31a7c7c
Author: efmer <boinc@efmer.eu>
Date:   Wed Oct 7 09:26:06 2020 +0200

    Make sure the authorized computer matches the IP and Port when approved

commit 6264409e95fdfa9a168efbe2ef467af184bc85b9
Author: efmer <boinc@efmer.eu>
Date:   Tue Oct 6 15:39:50 2020 +0200

    Handles switching the WiFi on and off.
    Better handling shutting down the connection thread

commit 61a0e83c52d321162087310d551b3f004f3691fe
Author: efmer <boinc@efmer.eu>
Date:   Mon Oct 5 15:16:24 2020 +0200

    Added Legacy MD5 password

commit 090ede1ea4e6efca2aa6d075217ce79b5cd3ed51
Merge: c794cf2d70 03b2e00990
Author: efmer <boinc@efmer.eu>
Date:   Mon Oct 5 11:57:35 2020 +0200

    Merge remote-tracking branch 'origin/master' into master

commit c794cf2d70c4411a8541ff36b2a10edc0b7f8278
Merge: 94b5214418 11af93eaf0
Author: efmer <boinc@efmer.eu>
Date:   Thu Oct 1 18:43:45 2020 +0200

    Merge remote-tracking branch 'origin/master' into master

commit 94b5214418c6c3970ad59be4a14348a3323876dd
Author: efmer <boinc@efmer.eu>
Date:   Mon Sep 28 12:40:15 2020 +0200

    Added Extern GUI (RpcExtern) Menu Fragment broadcast handlers to connection
    Not active yet: Password and secure connection

commit 7d368486fe1f8a4c77f1749c7a1ed9ad6352c993
Author: efmer <boinc@efmer.eu>
Date:   Tue Sep 22 18:49:08 2020 +0200

    External RPC

Fixes #

**Description of the Change**

**Alternate Designs**

**Release Notes**

### CODE REVIEW

**Comment by talregev** [comment]
Before I review, I want to ask to do some things:
1. Please provide a screenshots of the new GUI.
2. Please check that the new GUI it also work on old devices:
    a.k.a Run on AVD API 16 with arm support and crunch World Community Grid and Rosstea@Home.
    please provide a screenshot of the crunch.

**Comment by efmer** [comment]
As of November API 29 is mandatory for the Play Store.
But I did reset it to
        compileSdkVersion 29
        minSdkVersion 16
        targetSdkVersion 28
Otherwise it wouldn't start on my phone.

I can't test an older phone, but the Fragment isn't API dependent.

**Comment by efmer** [comment]
The screenshots
Screenshot_20201024-084305_BOINC
Screenshot_20201024-084332_BOINC

File: android/BOINC/app/build.gradle [hunk 1]
```diff
@@ -122,18 +122,20 @@ android {
...
+    dagger_version = '2.29.1'
     lifecycle_version = '2.2.0'
-    powermock_version = '2.0.5'
-    junit5_version = '5.6.2'
+    powermock_version = '2.0.7'
+    junit5_version = '5.7.0'
 }
 
 dependencies {
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/build.gradle:132]: Why'd you re-add this library?

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt [hunk 2]
```diff
@@ -203,11 +220,17 @@ class BOINCActivity : AppCompatActivity() {
         }
         super.onPause()
         unregisterReceiver(mClientStatusChangeRec)
+//        unregisterReceiver(mRpcExternBroadCastReceiver)
     }
 
     private fun doBindService() {
         // start service to allow setForeground later on...
-        startService(Intent(this, Monitor::class.java))
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt:228]: [`ContextCompat.startForegroundService()`](https://developer.android.com/reference/kotlin/androidx/core/content/ContextCompat#startforegroundservice) can be used here.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SplashActivity.java [hunk 3]
```diff
@@ -193,8 +193,16 @@ protected void onDestroy() {
     }
 
     private void doBindService() {
-        // start service to allow setForeground later on...
-        startService(new Intent(this, Monitor.class));
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/SplashActivity.java:196]: Same here.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt [hunk 4]
```diff
@@ -88,14 +102,19 @@ class Monitor : LifecycleService() {
     @Inject
     lateinit var noticeNotification: NoticeNotification
 
+    // External RPC
+    var mRpcExternServer = RpcExternServer()
+    var mRpcExternServerStart = false
+    private var manager: ConnectivityManager? = null
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:108]: Make this a lateinit variable, so a null check isn't needed every time.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt [hunk 5]
```diff
@@ -196,6 +216,40 @@ class Monitor : LifecycleService() {
...
+                override fun onLost(network: Network?) {
+                    if (!mRpcExternServerStart) return
+                    val wiFi = isconnectedToWifi()
+                    // not sure why we need a delay
+                    // but if we don't, rapidly switching the WiFi on / off doesn't work
+                    Thread.sleep(100)
+                    mRpcExternServer.update(wiFi, null)
+                }
+            }
+            manager = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager?
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:248]: Use the [`getSystemService`](https://developer.android.com/reference/kotlin/androidx/core/content/package-summary#(android.content.Context).getSystemService()) extension function here.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt [hunk 6]
```diff
@@ -224,6 +283,21 @@ class Monitor : LifecycleService() {
         } catch (e: Exception) {
             if (Logging.ERROR) Log.e(Logging.TAG, "Monitor.onDestroy error: ", e)
         }
+        stopForegroundService()
+    }
+
+    private fun stopForegroundService() {
+        // Stop foreground service and remove the notification.
+        stopForeground(true)
+        ContextCompat.getSystemService(this, NotificationManager::class.java)!!.cancelAll() // the above seems to fail
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:292]: Same here.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java [hunk 7]
```diff
@@ -272,6 +272,11 @@ public synchronized boolean authorize(String password) {
         return true;
     }
 
+
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java:275]: Why is there extra whitespace here?

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/authenticate/Hex.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package edu.berkeley.boinc.rpcExtern.authenticate
+
+class Hex {
+    val hexArray = "0123456789abcdef".toCharArray()
+    fun HexToByte(hexChars: String) : ByteArray {
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/authenticate/Hex.kt:5]: You could make these extension functions instead.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt [hunk 9]
```diff
@@ -38,17 +39,22 @@ import androidx.core.net.toUri
...
 import edu.berkeley.boinc.client.IMonitor
 import edu.berkeley.boinc.client.Monitor
 import edu.berkeley.boinc.databinding.MainBinding
+import edu.berkeley.boinc.rpcExtern.fragment.SettingsFragmentRpcExtern
 import edu.berkeley.boinc.ui.eventlog.EventLogActivity
 import edu.berkeley.boinc.utils.*
 import kotlinx.coroutines.launch
 
+
+// get Context using BOINCActivity.appContext
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt:56]: The application context can be accessed from any `Context` object, not just from activities.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExtern.kt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * Reads authentication token for GUI RPC authentication from file
+     *
+     * @param authFilePath absolute path to file containing GUI RPC authentication
+     * @return GUI RPC authentication code
+     */
+    fun readAuthToken(authFilePath: String): String {   //TODO eFMer duplicate ClientImplementation.java
+        var authKey = ""
+        try {
+            BufferedReader(FileReader(File(authFilePath))).use { br -> authKey = br.readLine() }
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExtern.kt:59]: You can call `File(authFilePath).bufferedReader()` instead.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExtern.kt [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            if (Logging.ERROR) {
+                Log.e(Logging.TAG, "Auth file not found: ", fnfe)
+            }
+        } catch (ioe: IOException) {
+            if (Logging.ERROR) {
+                Log.e(Logging.TAG, "IOException: ", ioe)
+            }
+        }
+        if (Logging.DEBUG) {
+            Log.d(Logging.TAG, "Authentication key acquired. length: " + StringUtils.length(authKey))
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExtern.kt:70]:
```suggestion
            Log.d(Logging.TAG, "Authentication key acquired. length: " + (authKey?.length ?: 0))
```

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt [hunk 12]
```diff
@@ -796,6 +919,54 @@ class Monitor : LifecycleService() {
         }
     }
 
+        private val mRpcExternBroadCastReceiver = RpcExternBroadcastReceiver()
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:922]: An anonymous class can be used instead (use `object : BroadcastReceiver()`).

**Comment by Isira-Seneviratne** [comment]
> As of November API 29 is mandatory for the Play Store.

In that case, the app architecture will have to be redone.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt [hunk 13]
```diff
@@ -247,8 +321,8 @@ class Monitor : LifecycleService() {
             val actionCode = intent.getIntExtra("action", -1)
             if (Logging.DEBUG) Log.d(Logging.TAG, "Monitor.onStartCommand() with action code: $actionCode")
             when (actionCode) {
-                1 -> lifecycleScope.launch { setClientRunMode(RUN_MODE_NEVER) }
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:250]: Why remove the space?

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExternString.kt [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package edu.berkeley.boinc.rpcExtern
+
+class RpcExternString {
+    val mRpcReplyBegin = "<boinc_gui_rpc_reply>"
+    val mRpcReplyEnd = "</boinc_gui_rpc_reply>"
+    val mRpcRequestBegin = "<boinc_gui_rpc_request>"
+}
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExternString.kt:7]:
```suggestion
const val mRpcReplyBegin = "<boinc_gui_rpc_reply>"
const val mRpcReplyEnd = "</boinc_gui_rpc_reply>"
const val mRpcRequestBegin = "<boinc_gui_rpc_request>"
```

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        data.externPasswrd = passwrd
+
+        try {
+            if (port.isNotEmpty()) {
+                data.externPort = port.toInt()
+            }
+            else
+            {
+                data.externPort = DEFAULT_PORT
+            }
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt:27]:
```suggestion
            data.externPort = if (port.isNotEmpty()) port.toInt() else DEFAULT_PORT
```

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+        if (data.ipAllowed3.isNotEmpty()) {
+            ipAllowedList.add(data.ipAllowed3)
+        }
+        if (data.ipAllowed4.isNotEmpty()) {
+            ipAllowedList.add(data.ipAllowed4)
+        }
+    }
+}
+
+class  RpcSettingsData() : Parcelable {
```
Comment (Isira-Seneviratne) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt:52]: The variables can be declared as part of the constructor.

File: android/BOINC/app/build.gradle (hunk 1 above)
Comment (efmer) [reply android/BOINC/app/build.gradle:132]: Sometimes things happen automatically.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt (hunk 2 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt:228]: Maybe but that version isn't really supporting it and might cause all kind of problems.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt (hunk 13 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:250]: Probably done automatically and why a space.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt (hunk 12 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:922]: Maybe, but this is this seems to be the preferred way

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java (hunk 7 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java:275]: Probably added something there to test and didn't remove the line

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExternString.kt (hunk 14 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcExternString.kt:7]: I  like classes

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt (hunk 15 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt:27]: I rather not, I do a lot of testing with the debugger and my way is easy the other is hard to debug.
Not much shorter, not easier to read, these are all personal things, not effecting the code.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt (hunk 16 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt:52]: I know but this worked more stable and the effect if probably similar, but hidden.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/authenticate/Hex.kt (hunk 8 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/authenticate/Hex.kt:5]: Maybe, but I liked to separate the new code. This is tested any change is not

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/SplashActivity.java (hunk 3 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/SplashActivity.java:196]: same here

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt (hunk 5 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:248]: ??

File: android/BOINC/app/build.gradle (hunk 1 above)
Comment (efmer) [reply android/BOINC/app/build.gradle:132]: I checked and do see a double, think it's a Kotlin legacy thing.
I removed it and will commit it

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt (hunk 9 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt]: Corrected

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt (hunk 2 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt:226]: The same thing is done other parts of the code.
I don't think this is needed in older versions. That said there are many changes in Services to keep it in the Foreground and it's still up to the phone to keep it running or kill it.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt (hunk 13 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.kt:250]: I can changed it, but am sure Android Studio will remove them again.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt (hunk 16 above)
Comment (efmer) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpcExtern/RpcSettingsData.kt:46]: To clarify I used the single Parcelize method, but this gave erratic results, this works fine.

**Review by efmer — COMMENTED** [review]
For now I suggest to leave most of the new files as they are.
Unless it's obviously wrong.
Some functions are picky and hardware related.

**Comment by talregev** [comment]
> As of November API 29 is mandatory for the Play Store.
> But I did reset it to
> compileSdkVersion 29
> minSdkVersion 16
> targetSdkVersion 28
> Otherwise it wouldn't start on my phone.
> 
> I can't test an older phone, but the Fragment isn't API dependent.

We know that API 29 in google play will be mandatory. It a problem we need to deal in other tread. 
My colleague here think the problem to switch API 29 is connected to architecture design. 
I think it connected the way we run the process. (Maybe to run it on foreground like you try but doing something more).
But again, this topic is not for this PR and you should revert all the connected API 29 change including the foreground change you made here:
https://github.com/BOINC/boinc/pull/4065/files#diff-4d321c252413c45ecb07c95debe75ad0ed1c25a5cd8f9c98ea3a308bbe1d57d5L197

You can check the GUI on old device. And I will not approve my review until you do it.
If you use Android Studio, there is AVD = Android Virtual Device. It an emulator aldo for old devices.
You should download arm with google api API 16 emulator image (do it with google AVD), then run on the emulator this gui including crunch with calculatio. 
Don't say it just a GUI and it not affect the calculation. We change the gui here couple time and it make some crush and other things that interfere with the crunch. Also things looks great on new device, not so on old device, that why you need to do this check. 

We also need to check this PR in security aspect. I am not the qualify person to review this aspect, maybe @AenBleidd  Can add more observation in this matter. 

Also thank you for the screenshot for your PR.
Also can you give a screenshot how external GUI look like and how to connect? 

Thank you for your hard work.
I hope you understand it constructive comments that will  increase the quality of your PR.

**Comment by efmer** [comment]
In my experience the AVD isn't a very reliable testing tool. I hardly ever use it, the actual phone often gives other results, things that work perfectly on the AVD doesn't work on an actual device.
Supporting older devices is a crime, I'm not that sure that even Google tests much on API 16.
I want to make too many changes and have seen many more things I don't like so I just build a new app from the ground up.
Splitting the Client from the GUI, like I've done makes things simpler and more stable.

**Comment by talregev** [comment]
> In my experience the AVD isn't a very reliable testing tool. I hardly ever use it, the actual phone often gives other results, things that work perfectly on the AVD doesn't work on an actual device.
> Supporting older devices is a crime, I'm not that sure that even Google tests much on API 16.
> I want to make too many changes and have seen many more things I don't like so I just build a new app from the ground up.
> Splitting the Client from the GUI, like I've done makes things simpler and more stable.

Not sure what you want to say.
We already change the gui of the app and have some experience with GUI also with AVD and old phones.
AVD is a great tool to detect and fix problem. I don't say it perfect but it better then nothing. 
You don't have physical old device and you don't want to use AVD to check your GUI.
I will not go forward with this plan.

**Pull request closed without merging**